builders/gradle.go

```diff
@@ -0,0 +1,149 @@
+package builders
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	logging "github.com/op/go-logging"
+
+	"github.com/fossas/fossa-cli/config"
+	"github.com/fossas/fossa-cli/module"
+)
+
+var gradleLogger = logging.MustGetLogger("gradle")
+
+// GradleBuilder implements Builder for build.gradle builds
+type GradleBuilder struct {
+	GradleCmd     string
+	GradleVersion string
+}
+
+// Initialize collects metadata on Java and Maven binaries
+func (builder *GradleBuilder) Initialize() error {
+	gradleLogger.Debug("Initializing Gradle builder...")
+
+	// Set Gradle context variables
+	gradleCmd, gradleVersionOut, err := which("--version --offline", os.Getenv("GRADLE_BINARY"), "./gradlew", "gradle")
+	if err == nil {
+		builder.GradleCmd = gradleCmd
+
+		gradleVersionMatchRe := regexp.MustCompile(`Gradle ([0-9]+\.[0-9]+.\w+)`)
+		match := gradleVersionMatchRe.FindStringSubmatch(gradleVersionOut)
+		if len(match) == 2 {
+			builder.GradleVersion = match[1]
+		}
+	}
+
+	if builder.GradleCmd == "" {
+		return fmt.Errorf("could not find Gradle binary or wrapper (try setting $GRADLE_BINARY)")
+	}
+
+	gradleLogger.Debugf("Done initializing Gradle builder: %#v", builder)
+	return nil
+}
+
+// Build is not implemented
+func (builder *GradleBuilder) Build(m module.Module, force bool) error {
+	return nil
+}
+
+// Analyze parses the output of `gradle -q app:dependencies`
+func (builder *GradleBuilder) Analyze(m module.Module, allowUnresolved bool) ([]module.Dependency, error) {
+	gradleLogger.Debugf("Running Gradle analysis: %#v %#v in %s", m, allowUnresolved, m.Dir)
+
+	// TODO: We need to let the user configure the right configurations
+	dependenciesOutput, err := exec.Command(builder.GradleCmd, m.Name+":dependencies", "-q", "--configuration=compile", "--offline", "-a").Output()
+	// Do not handle error, as gradle will exit with status 1 regardless
+	if len(dependenciesOutput) == 0 || err != nil {
+		return nil, fmt.Errorf("could not run Gradle task %s:dependencies", m.Name)
+	}
+
+	var deps []module.Dependency
+	dependenciesRe := regexp.MustCompile(`- ([\w\.-]+):([\w\.-]+):([\w\.-]+)( -> ([\w\.-]+))?`)
```

Comment:
This regex is unmaintainable without a very detailed comment. I think it would be much cleaner to loop through each line and check for `---` instead (which prefixes all lines with dependency names).

Reply:
That's not true, since it renders as a tree the parent nodes have a different prefix.  There is no way to guarantee a specific format for each level as well.

The only common prefix i found was `- `

Reply:
Really? I see all nodes being prefixed with either `\--- ` or `+--- `. Is this because we're using different versions? Do you have an example project where this isn't true?